lib/evmone/tracing.cpp

```diff
@@ -0,0 +1,59 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2021 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "tracing.hpp"
+#include <evmc/hex.hpp>
+
+namespace evmone
+{
+namespace
+{
+/// @see create_histogram_tracer()
+class HistogramTracer : public Tracer
+{
+    evmc_revision m_rev{};
+    int32_t m_depth = 0;
+    const uint8_t* m_code = nullptr;
+    uint32_t m_counts[256]{};
+    std::ostream& m_out;
+
+    void on_execution_start(
+        evmc_revision rev, const evmc_message& msg, bytes_view code) noexcept override
+    {
+        m_rev = rev;
+        m_depth = msg.depth;
+        m_code = code.data();
+    }
+
+    void on_instruction_start(uint32_t pc) noexcept override { ++m_counts[m_code[pc]]; }
+
+    void on_execution_end(const evmc_result& /*result*/) noexcept override
+    {
+        const auto names = evmc_get_instruction_names_table(m_rev);
+        m_out << "--- # HISTOGRAM depth=" << m_depth << "\nopcode,count\n";
+        for (size_t i = 0; i < std::size(m_counts); ++i)
+        {
+            if (m_counts[i] != 0)
+            {
+                if (const auto name = names[i]; name != nullptr)
+                    m_out << name;
+                else
+                    m_out << "0x" << evmc::hex(static_cast<uint8_t>(i));
+                m_out << ',' << m_counts[i] << '\n';
+            }
+        }
+
+        m_code = nullptr;
```

Comment:
Maybe zero out `m_counts` and `m_depth` then, too?

Reply:
True.

Reply:
Well... because we have only single table in tracer we cannot measure internal calls separately. Therefore we should output only at depth=0 (and only then zeroing counters make sense). Moreover, reporting depth is useless.

Reply:
Wait, is `on_execution_start` called for each subcall? It changes `m_code`, how would it work after return from subcall then?

Reply:
Oh shit, it was suppose to be easy tracer example...